spark/src/main/java/org/apache/iceberg/actions/BaseSparkAction.java

```diff
@@ -50,6 +53,24 @@
 
   protected abstract Table table();
 
+  protected abstract R doExecute();
+
+  protected abstract JobGroupInfo jobGroup();
```

Comment:
As much as I like having this in a single place, I am not sure this will work for actions that trigger multiple jobs. For example, data compaction @RussellSpitzer is working on will submit a job per partition and we would want to a separate description each time.

I've just merged a refactoring of `BaseSparkAction` that now makes SparkContext accessible in this class.

I think we can have this:

```
  protected <T> T withJobGroupInfo(JobGroupInfo info, Supplier<R> supplier) {
    JobGroupInfo previousInfo = JobGroupUtils.getJobGroupInfo(sparkContext);
    try {
      JobGroupUtils.setJobGroupInfo(sparkContext, info);
      return supplier.get();
    } finally {
      JobGroupUtils.setJobGroupInfo(sparkContext, previousInfo);
    }
  }
```

Then, yes, we would need to modify more code in the existing actions but it should not be that bad.

```
public ... execute() {
  JobGroupInfo info = ... 
  withJobGroupInfo(info, () -> {
    // call doExecute (which is now points to the code in execute)
  })
}
```

Reply:
I'd like to hear @RussellSpitzer's thoughts on this.

Reply:
That would be helpful for me, I will be having various actions fire off an will need multiple descriptions. In my future thoughts for distributed planning we would similarly want something that can apply separate job group info to separate actions.

Too bad this isn't Scala it would be pretty :)

Reply:
`withJobGroupInfo` is starting to look more like a try-with-resource statement. I think that would be a more java-y way of expressing that. And to @RussellSpitzer's point it wouldn't require any icky Scala :stuck_out_tongue:

Reply:
@rymurr, do you mean making `JobGroupInfo` implement `AutoCloseable` and setting the new job description in the constructor?

```
  public ... execute() {
    try (JobGroupInfo ignored = newJobGroupInfo(...)) {
      doExecute();
    }
  }
```

If yes, both approaches work for me, I have no preference.

Reply:
Thought, try-with-resources makes it shorter. Maybe, that's a better option.